[futures] Add c++ ivalue::Future collectAll() helper

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39597 [futures] Add collectAny() to ivalue::Future for completeness
* **#39119 [futures] Add c++ ivalue::Future collectAll() helper**

Add some base c++ unittest coverage for ivalue::Future, and in
the process, add a basic collectAll() primitive, per https://github.com/pytorch/pytorch/issues/38937.

In the process, I realized that List<Future> is effectively
impossible to construct (since the Future's type is not templated,
but rather passed in,  the getTypePtr_<T>::call() isn't defined),
so added a workaround in List to make it possible.

Differential Revision: [D21756884](https://our.internmc.facebook.com/intern/diff/D21756884/)

Commit:
[futures] Add c++ ivalue::Future collectAll() helper

Add some base c++ unittest coverage for ivalue::Future, and in
the process, add a basic collectAll() primitive, per 38937.

In the process, I realized that List<Future> is effectively
impossible to construct (since the Future's type is not templated,
but rather passed in,  the getTypePtr_<T>::call() isn't defined),
so added a workaround in List to make it possible.

Differential Revision: [D21756884](https://our.internmc.facebook.com/intern/diff/D21756884/)

[ghstack-poisoned]

Commit:
Update on "[futures] Add c++ ivalue::Future collectAll() helper"


Add some base c++ unittest coverage for ivalue::Future, and in
the process, add a basic collectAll() primitive, per https://github.com/pytorch/pytorch/issues/38937.

In the process, I realized that List<Future> is effectively
impossible to construct (since the Future's type is not templated,
but rather passed in,  the getTypePtr_<T>::call() isn't defined),
so added a workaround in List to make it possible.

Differential Revision: [D21756884](https://our.internmc.facebook.com/intern/diff/D21756884/)

[ghstack-poisoned]

Commit:
clang-format on "[futures] Add c++ ivalue::Future collectAll() helper"


Add some base c++ unittest coverage for ivalue::Future, and in
the process, add a basic collectAll() primitive, per https://github.com/pytorch/pytorch/issues/38937.

In the process, I realized that List<Future> is effectively
impossible to construct (since the Future's type is not templated,
but rather passed in,  the getTypePtr_<T>::call() isn't defined),
so added a workaround in List to make it possible.

Differential Revision: [D21756884](https://our.internmc.facebook.com/intern/diff/D21756884/)

[ghstack-poisoned]

Commit:
oss-only export issue on "[futures] Add c++ ivalue::Future collectAll() helper"


Add some base c++ unittest coverage for ivalue::Future, and in
the process, add a basic collectAll() primitive, per https://github.com/pytorch/pytorch/issues/38937.

In the process, I realized that List<Future> is effectively
impossible to construct (since the Future's type is not templated,
but rather passed in,  the getTypePtr_<T>::call() isn't defined),
so added a workaround in List to make it possible.

Differential Revision: [D21756884](https://our.internmc.facebook.com/intern/diff/D21756884/)

[ghstack-poisoned]

mrshenli commented on aten/src/ATen/core/List_inl.h:
```diff
@@ -43,7 +43,8 @@ List<T>::List(TypePtr elementType)
     typename c10::detail::ListImpl::list_type(),
     std::move(elementType))) {
-  static_assert(std::is_same<T, IValue>::value, "This constructor is only valid for c10::impl::GenericList.");
+  static_assert(std::is_same<T, IValue>::value || std::is_same<T, c10::intrusive_ptr<ivalue::Future>>::value,
```
@wanchaol is this change OK?

jjlilley replied:
Oh, and to clarify this one - the issue is that `ivalue::Future` doesn't (and maybe can't trivially) define a `getTypePtr_::call()` operator to get its type, since the type isn't part of the template.

Without  an exception to the assert, it's not possible to have a `c10::List` that stores objects that don't have the `call()` operator defined.

jjlilley replied:
@wanchaol Could you take a look, particularly this line?

wanchaol replied:
hmmm sorry for the late check. I am wondering why we need this check, in particular, `GenericList` already stores a list of ivalues, including the possible ivalue::Future (of course you can't directly store c10::List<ivalue::Future>, but we can use c10::List<IValue> already, where the IValue can be turned into ivalue::Future by `ivalue.toFuture`.

jjlilley replied:
Yes, it is possible to instantiate a `GenericList` of generic `IValues` (though I think the api is technically internal only?)

For this api, I was hoping to make the api slightly more typed, i.e. `List<Future>`

That said, without this one-liner, it's simply not possible to instantiate a `List<Future>` (as it is possible to instantiate `List<Tensor>`, `List<int64_t>`, etc) without a compile error. This part seems kind of odd from an api consistency standpoint.

Here's a concrete code snippet:
```
   // Future to insert
    auto f1 = make_intrusive<Future>(IntType::get());                                                                                    
    // Untyped list works just fine
    c10::impl::GenericList genericList(AnyType::get());                                                                                       
    genericList.push_back(f1);                                                                                                                

    // This will not compile without the 1-liner above, because of the lack of a 
   // definition (or good way of doing so) of the "call()" operator for Futures
    List<intrusive_ptr<ivalue::Future>> typedList(FutureType::create(AnyType::get()));                                                   
    typedList.push_back(f1);                                                                                                                  
```

I think it would be nice to be able to create List<Future> on occasion when it's desired... that's all that this line attempts to do.

mrshenli commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -604,4 +604,45 @@ getClassConverter() {
+    ctx->srcFutures.get(i)->addCallback(func);
+  }
+  if (ctx->srcFutures.size() == 0) {
+    ctx->dstFuture->markCompleted(ctx->asIvalue);
```
The IValue type of the `dstFuture` is a list of futures? And this is because we wanted to conveniently retrieve all results from the `dstFuture`?

jjlilley replied:
Not sure if there's a better suggestion?

It seems that the result should be a Future of some sort. It seems like the second-best alternative would be a Future<Null> where the called needs to remember the passed in futures out-of-band? But this didn't necessarily seem better...

A `Future<List<IValue>>` isn't really possible because of the need to propagate exceptions back. which is why it was effectively `Future<List<Future<IValue>>>`

mrshenli replied:
I see. This current version LGTM.

mrshenli commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -604,4 +604,45 @@ getClassConverter() {
+CAFFE2_API intrusive_ptr<ivalue::Future> collectAll(
+    std::vector<intrusive_ptr<ivalue::Future>> srcs) {
+  auto typePtr =
+      !srcs.empty() ? srcs[0]->type() : FutureType::create(NoneType::get());
```
This means all `Future`s in `srcs` need to be the same type? Do we need to `TORCH_CHECK` that?

jjlilley replied:
Yes, this is effectively (and surprisingly) an artifact of the current `List<>` and `Future<>` apis.
I'll add a `TORCH_CHECK()`.

`c10::List` requires every entry to have the same underlying type, and `Future<Any_IValue>` isn't quite possible - a specific `TypePtr` is required. I was a bit surprised by some of the constraints imposed by the IValue library.

mrshenli replied:
(not requesting changes to this PR) I wonder if it is possible to support collecting a list of Futures with different types?

jjlilley replied:
actually, I didn't realize that there was `AnyType` in `jit_type.h`
If I change typeptr to `FutureType::create(AnyType::get());` things seem to work, and I don't need any of these checks, which is much better.

jjlilley replied:
update (also in comment in other area) is that I realized that ` FutureType::create(AnyType::get())` seems to work, which is better

mrshenli reviewed (commented):
Lint failures look relevant, but it does not seem to print the details. Maybe try running `clang-format` locally? 

```
diff --git a/test/cpp/jit/test_misc.cpp b/test/cpp/jit/test_misc.cpp index 15c491e064..192f20b3f8 100644 --- a/test/cpp/jit/test_misc.cpp +++ b/test/cpp/jit/test_misc.cpp @@ -1311,8 +1311,10 @@ graph(%a): } } -static void -checkShape(Node* n, std::vector<int64_t> expected, bool prev = true) { +static void checkShape( + Node* n, + std::vector<int64_t> expected, + bool prev = true) { auto profile = (prev) ? n->inputs().at(0)->node() : n; auto tp = profile->output()->type(); auto ptp = tp->expect<TensorType>();
```

mrshenli commented:
Requesting review from @wanchaol for JIT-related changes.

Commit:
Update on "[futures] Add c++ ivalue::Future collectAll() helper"


Add some base c++ unittest coverage for ivalue::Future, and in
the process, add a basic collectAll() primitive, per https://github.com/pytorch/pytorch/issues/38937.

In the process, I realized that List<Future> is effectively
impossible to construct (since the Future's type is not templated,
but rather passed in,  the getTypePtr_<T>::call() isn't defined),
so added a workaround in List to make it possible.

Differential Revision: [D21756884](https://our.internmc.facebook.com/intern/diff/D21756884/)

[ghstack-poisoned]

jjlilley commented:
Re: clang-format, the test_misc.cpp is related to pre-existing tests, which I had clang-formatted but in the followup selectively reverted.

btw, is there a tip for running clang-format within fbcode/caffe2? I have a somewhat indirect/hacky mechanism for feeding to clang-format, though realizing that the params I'm using might be slightly different. The caffe2/tools/clang_tidy.py scripts don't work within fbcode (i.e. they assume git, python2, etc) - I suppose I can look at this, though there's been some inertia as my own hacky custom scripts for this have worked mostly well enough...

Commit:
Update on "[futures] Add c++ ivalue::Future collectAll() helper"


Add some base c++ unittest coverage for ivalue::Future, and in
the process, add a basic collectAll() primitive, per https://github.com/pytorch/pytorch/issues/38937.

In the process, I realized that List<Future> is effectively
impossible to construct (since the Future's type is not templated,
but rather passed in,  the getTypePtr_<T>::call() isn't defined),
so added a workaround in List to make it possible.

Differential Revision: [D21756884](https://our.internmc.facebook.com/intern/diff/D21756884/)

[ghstack-poisoned]

mrshenli commented:
> btw, is there a tip for running clang-format within fbcode/caffe2? 

I usually just use `cd some/folder; clang-format -i *.h *.cpp` on devservers. The clang-format installed there seems to satisfy internal and OSS requirements.

mrshenli reviewed (approved):
LGTM!

Commit:
rebase on "[futures] Add c++ ivalue::Future collectAll() helper"


Add some base c++ unittest coverage for ivalue::Future, and in
the process, add a basic collectAll() primitive, per https://github.com/pytorch/pytorch/issues/38937.

In the process, I realized that List<Future> is effectively
impossible to construct (since the Future's type is not templated,
but rather passed in,  the getTypePtr_<T>::call() isn't defined),
so added a workaround in List to make it possible.

Differential Revision: [D21756884](https://our.internmc.facebook.com/intern/diff/D21756884/)

[ghstack-poisoned]

wanchaol commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -412,6 +412,13 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
 
+// Input is a list of Futures with the same target type.
+// Output is a Future to the List of completed Futures.
+CAFFE2_API intrusive_ptr<ivalue::Future> collectAll(
```
Ok it seems like this is where we want the c10::List to have ivalue::Future, but why we have two APIs here? shouldn't the second one be enough?

jjlilley replied:
My intent was to just create a convenience api (just like some of the alternate constructor apis in ivalue.h), since List<> can entail some effort to use (though maybe just me? :) ). I can remove if we'd really like.

jjlilley replied:
Actually, looking at this a bit more, I'll remove this secondary api.

jjlilley replied:
Actually, un-removing after hacking on the unittest a bit, unless there's an objection.

wanchaol replied:
How would this c++ API getting used in the rpc? I don't have strong opinion on if we should use c10::List as a C++ API for this, but one drawback I am seeing here is, c10::List is suppose to holding a GenericList with the same type (acting like a real list that contains a single type), but if we expose this c10::List with Future as a exception on the `static_assert` we discussed above, we might ended up having something like `List[Future[int], Future[double]]`, which is not desirable as it break the list contraints in the type system, which might potentially cause some issue.

jjlilley replied:
Re: rpc - since rpc exported api is based on `torch::util::Future<Message>` (since `Message` isn't an `IValue`), it should be straightforward to create a parallel version for the `torch::util::Future` type.

Re: `c10::List<>` - regardless of the input (`vector` vs `List`), we need to output a `c10::List`, because it needs to be the object of a `Future<>` (i.e. `Future<List<Future<Any>>>`. If we use `type FutureType(AnyType)` in the `c10::List`, things do seem to work (and not like `c10::List` enforces this for `Future` - remember the lack of the `call` operator, or type firmly embedded in the template)

wanchaol replied:
What I am trying to see for this API is that, are we having the case where we want `collectAll` takes a list like `List[Future[int], Future[double], Future[Tensor]`? Or should `collectAll` be only applies to a list that have the same type? 

If the use case is the later one, we should just use the `std::vector` for all cases (c10::List version can be used internally as implementation to generate the new Future)? at least this seems to be the one used by C++ tests natively. We should also create the `Future<List<Future<TypeFromInput>>>` instead of AnyType to make the type consistent before and after.

Commit:
remove vector-based secondary api on "[futures] Add c++ ivalue::Future collectAll() helper"


Add some base c++ unittest coverage for ivalue::Future, and in
the process, add a basic collectAll() primitive, per https://github.com/pytorch/pytorch/issues/38937.

In the process, I realized that List<Future> is effectively
impossible to construct (since the Future's type is not templated,
but rather passed in,  the getTypePtr_<T>::call() isn't defined),
so added a workaround in List to make it possible.

Differential Revision: [D21756884](https://our.internmc.facebook.com/intern/diff/D21756884/)

[ghstack-poisoned]

Commit:
un-remove vector-based secondary api on "[futures] Add c++ ivalue::Future collectAll() helper"


Add some base c++ unittest coverage for ivalue::Future, and in
the process, add a basic collectAll() primitive, per https://github.com/pytorch/pytorch/issues/38937.

In the process, I realized that List<Future> is effectively
impossible to construct (since the Future's type is not templated,
but rather passed in,  the getTypePtr_<T>::call() isn't defined),
so added a workaround in List to make it possible.

Differential Revision: [D21756884](https://our.internmc.facebook.com/intern/diff/D21756884/)

[ghstack-poisoned]

wanchaol commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -604,4 +604,43 @@ getClassConverter() {
+
+CAFFE2_API intrusive_ptr<ivalue::Future> collectAll(
+    std::vector<intrusive_ptr<ivalue::Future>> srcs) {
+  List<intrusive_ptr<ivalue::Future>> asList(FutureType::create(AnyType::get()));
```
we should create the List of Future with type as the type inside `srcs` i think? e.g. FutureType::create(srcs[0].type())

jjlilley replied:
How is that better than this approach? Is it set in stone that all Future target types must be the same?

wanchaol replied:
Yes, this also make sure the return type remains accurate according to its input future_list, so that adding torchscript support for this API is valid that we have accurate type information from the output of this API.

jjlilley replied:
Sure, that's fine.
 - The `c10::List` version of the api was already type-preserving, so no issues there.
 - Updating the vector version to use the appropriate type if they all match.
 - In the process, I found that `ivalue::Future`'s handling of typing is a bit inconsistent - `Future:type()` was like `List::elementType()` than e.g. the `Tuple::type()`. I ended up renaming this, and we can introduce a separate `Future::type()` if necessary.

Commit:
update based on Wanchao's typing comments on "[futures] Add c++ ivalue::Future collectAll() helper"


Add some base c++ unittest coverage for ivalue::Future, and in
the process, add a basic collectAll() primitive, per https://github.com/pytorch/pytorch/issues/38937.

In the process, I realized that List<Future> is effectively
impossible to construct (since the Future's type is not templated,
but rather passed in,  the getTypePtr_<T>::call() isn't defined),
so added a workaround in List to make it possible.

Differential Revision: [D21756884](https://our.internmc.facebook.com/intern/diff/D21756884/)

[ghstack-poisoned]

wanchaol commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -604,4 +604,55 @@ getClassConverter() {
+    TypePtr res = srcs[0]->elementType();
+    for (size_t i = 1, len = srcs.size(); i < len; ++i) {
+      if (*srcs[i]->elementType() != *res) {
+        return FutureType::create(AnyType::get());
```
I guess if there's no use case that takes multiple future of different types into a single list, we should error out to the user in this case rather than use the AnyType?

jjlilley replied:
There is no current use case, though if it's simple to enable, not clear why it needs to be ane error?

wanchaol replied:
I feel like this will break the list type constraint, e.g. a list like `List{Future[Int], Future[Double]}` is wrong in theory. For a single list instance, we should always have the same `element_type` for all the elements, so user have to construct two list with Future[Int] and Future[Double] separately and call `collectAll` separately, but we seems enabling this wrong pattern in this PR.

jjlilley replied:
That's fine, I'll add a `TORCH_CHECK()`

Commit:
add TORCH_CHECK for consistent List typing on "[futures] Add c++ ivalue::Future collectAll() helper"


Add some base c++ unittest coverage for ivalue::Future, and in
the process, add a basic collectAll() primitive, per https://github.com/pytorch/pytorch/issues/38937.

In the process, I realized that List<Future> is effectively
impossible to construct (since the Future's type is not templated,
but rather passed in,  the getTypePtr_<T>::call() isn't defined),
so added a workaround in List to make it possible.

Differential Revision: [D21756884](https://our.internmc.facebook.com/intern/diff/D21756884/)

[ghstack-poisoned]

wanchaol commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -604,4 +604,49 @@ getClassConverter() {
+  // Lists are generally expected to have a consistent type, check
+  // for this unless we have a compelling target use case.
+  for (size_t i = 1, len = srcs.size(); i < len; ++i) {
+    TORCH_CHECK(*srcs[i]->elementType() == *typePtr);
```
Maybe add a error message like `"Expecting type " typePtr->python_str() " in the list but found type " srcs[i]->elementType()->python_str()` ?

wanchaol reviewed (approved):
Looks good! Thanks for addressing the typing issue.

Commit:
Update on "[futures] Add c++ ivalue::Future collectAll() helper"


Add some base c++ unittest coverage for ivalue::Future, and in
the process, add a basic collectAll() primitive, per https://github.com/pytorch/pytorch/issues/38937.

In the process, I realized that List<Future> is effectively
impossible to construct (since the Future's type is not templated,
but rather passed in,  the getTypePtr_<T>::call() isn't defined),
so added a workaround in List to make it possible.

Differential Revision: [D21756884](https://our.internmc.facebook.com/intern/diff/D21756884/)

[ghstack-poisoned]
